ocamlPackages.smtml: init at 0.7.0

ethancedwards8 opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md), [pkgs/README.md](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md), [maintainers/README.md](https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md) and other contributing documentation in corresponding paths.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

ethancedwards8 committed:
ocamlPackages.smtml: init at 0.7.0

Signed-off-by: Ethan Carter Edwards <ethan@ethancedwards.com>

ethancedwards8 commented on pkgs/development/ocaml-modules/smtml/default.nix:
```diff
@@ -0,0 +1,78 @@
+  ];
+
+  # Error: Library "mdx.top" not found.
+  doCheck = false;
```
@vbgl any thoughts on how to troubleshoot this mdx error?

ethancedwards8 commented:
New error!
```
Ran: 16 tests in: 0.71 seconds.
OK
File "dune", lines 1-8, characters 0-199:
1 | (mdx
2 |  (enabled_if
3 |   (and
4 |    %{lib-available:z3}
5 |    (not %{lib-available:alt-ergo-lib})
6 |    (not %{lib-available:bitwuzla-cxx})
7 |    (not %{lib-available:colibri2.core})
8 |    (not %{lib-available:cvc5}))))
(cd _build/default && ./mdx_gen.bc.exe README.md) > _build/default/.mdx/README.md.corrected
Got an error while evaluating:
---
let stats = Z3.get_statistics solver;;
---
Line 1, characters 13-30:
Error: Unbound value Z3.get_statistics
File "doc/dune", lines 4-15, characters 0-288:
 4 | (mdx
 5 |  (deps
 6 |   (source_tree examples)
....
13 |    (not %{lib-available:colibri2.core})
14 |    (not %{lib-available:cvc5})))
15 |  (files :standard - *.mld))
(cd _build/default/doc && ./mdx_gen.bc.exe index.mld) > _build/default/doc/.mdx/index.mld.corrected
Got an error while evaluating:
---
Z3.get_statistics solver
---
```
Is this an incompatibility between z3 and smtml tests?

vbgl commented:
Strange. CI builds are successful so far…

vbgl commented:
This is as successful build: https://logs.ofborg.org/?key=nixos/nixpkgs.418547&attempt_id=46c7f88f-8767-4770-a12c-f846343c6eb7

ethancedwards8 commented:
Looks like it fails on Darwin: https://github.com/ethancedwards8/nixpkgs-review-gha/actions/runs/15829327941

I may just disable Darwin for the lib?

ethancedwards8 commented:
Yeah failed nix build. But when I build on my local Darwin machines it fails but on Linux it passes.

ethancedwards8 commented:
> New error!
> 
> ```
> Ran: 16 tests in: 0.71 seconds.
> OK
> File "dune", lines 1-8, characters 0-199:
> 1 | (mdx
> 2 |  (enabled_if
> 3 |   (and
> 4 |    %{lib-available:z3}
> 5 |    (not %{lib-available:alt-ergo-lib})
> 6 |    (not %{lib-available:bitwuzla-cxx})
> 7 |    (not %{lib-available:colibri2.core})
> 8 |    (not %{lib-available:cvc5}))))
> (cd _build/default && ./mdx_gen.bc.exe README.md) > _build/default/.mdx/README.md.corrected
> Got an error while evaluating:
> ---
> let stats = Z3.get_statistics solver;;
> ---
> Line 1, characters 13-30:
> Error: Unbound value Z3.get_statistics
> File "doc/dune", lines 4-15, characters 0-288:
>  4 | (mdx
>  5 |  (deps
>  6 |   (source_tree examples)
> ....
> 13 |    (not %{lib-available:colibri2.core})
> 14 |    (not %{lib-available:cvc5})))
> 15 |  (files :standard - *.mld))
> (cd _build/default/doc && ./mdx_gen.bc.exe index.mld) > _build/default/doc/.mdx/index.mld.corrected
> Got an error while evaluating:
> ---
> Z3.get_statistics solver
> ---
> ```
> 
> Is this an incompatibility between z3 and smtml tests?

Darwin fails with this error. For my uses I only need it to work on Linux so I am okay with marking it broken on Darwin.

ethancedwards8 commented on pkgs/development/ocaml-modules/smtml/default.nix:
```diff
@@ -0,0 +1,80 @@
+    downloadPage = "https://github.com/formalsec/smtml";
+    changelog = "https://github.com/formalsec/smtml/releases/tag/v${version}";
+    license = lib.licenses.mit;
+    broken = stdenv.hostPlatform.isDarwin;
```
The log I posted as text (not the link) was a full build log on Darwin. Thanks.

ethancedwards8 commented on pkgs/development/ocaml-modules/smtml/default.nix:
```diff
@@ -0,0 +1,80 @@
+    sedlex
+    zarith
+  ];
+
```
Fixed.

ethancedwards8 commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 418547 --package ocamlPackages.smtml`
Commit: `9a0046e2ac9c8e0075c16544f42c68fa68a88895`

---
### `aarch64-darwin`
<details>
  <summary>:x: 1 package failed to build:</summary>
  <ul>
    <li>ocamlPackages.smtml</li>
  </ul>
</details>

---
<details>
<summary>Error logs: `aarch64-darwin`</summary>
<details>
<summary>ocamlPackages.smtml</summary>
<pre>                                               
File "doc/dune", lines 4-15, characters 0-288:
 4 | (mdx
 5 |  (deps
 6 |   (source_tree examples)
....
13 |    (not %{lib-available:colibri2.core})
14 |    (not %{lib-available:cvc5})))
15 |  (files :standard - *.mld))
(cd _build/default/doc && ./mdx_gen.bc.exe index.mld) > _build/default/doc/.mdx/index.mld.corrected
Got an error while evaluating:
---
Z3.get_statistics solver
---
Line 1, characters 1-18:
Error: Unbound value Z3.get_statistics
Done: 99% (543/546, 3 left, 1 failed) (jobs: 2)
                                               
Done: 99% (543/546, 3 left, 2 failed) (jobs: 1)
                                               </pre>
</details>
</details>

ethancedwards8 commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review-gha`](https://github.com/Defelo/nixpkgs-review-gha)

Command: `nixpkgs-review pr 418547`

Logs: https://github.com/ethancedwards8/nixpkgs-review-gha/actions/runs/15853834974

vbgl commented:
The error is very strange. You may want to disable tests on darwin since the library seems to build fine otherwise.

vbgl commented:
@ofborg build ocaml-ng.ocamlPackages_5_0.smtml

vbgl commented on pkgs/development/ocaml-modules/smtml/default.nix:
```diff
@@ -0,0 +1,74 @@
+    mdx.bin
+  ];
+
+  doCheck = !stdenv.hostPlatform.isDarwin;
```
Tests fail for OCaml 5.0 so you may want to also disable them in that case.

ethancedwards8 commented on pkgs/development/ocaml-modules/smtml/default.nix:
```diff
@@ -0,0 +1,74 @@
+  fpath,
+  hc,
+  menhirLib,
+  ocaml_intrinsics,
```
Nice.
